packages/beacon-node/src/chain/chain.ts

```diff
@@ -934,6 +937,42 @@ export class BeaconChain implements IBeaconChain {
     if (headState) {
       this.opPool.pruneAll(headBlock, headState);
     }
+
+    const cpEpoch = cp.epoch;
+    const eip6110Epoch = headState?.config.EIP6110_FORK_EPOCH ?? Infinity;
+
+    if (headState === null) {
+      this.logger.verbose("Head state is null");
+    } else if (cpEpoch >= eip6110Epoch) {
+      const headEpoch = headState.epochCtx.epoch;
+
+      // Get the validator.length from the state at cpEpoch
+      // We are confident the last element in the list is from headEpoch
+      // Thus we query from the end of the list. (cpEpoch - headEpoch - 1) is negative number
+      const pivotValidatorIndex = headState.epochCtx.historicalValidatorLengths.get(cpEpoch - headEpoch - 1);
+
+      if (pivotValidatorIndex !== undefined) {
+        // Note EIP-6914 will break this logic
+        const newFinalizedValidators = headState.epochCtx.unfinalizedPubkey2index.filter(
+          (index, _pubkey) => index < pivotValidatorIndex
+        );
+
+        // Populate finalized pubkey cache and remove unfinalized pubkey cache
+        if (!newFinalizedValidators.isEmpty()) {
+          this.regen.updateUnfinalizedPubkeys(newFinalizedValidators);
+        }
+      }
+    }
+
+    // TODO-6110: Deprecating eth1Data poll requires a check on a finalized checkpoint state.
```

Comment:
@dapplion eth1data poll deprecation requires `finalizedState` which is recently removed. Will follow up in another PR when this is resolved. The lack of deprecation shouldn't impact the core functionality and performance of 6110.